File: 117. Populating Next Right Pointers in Each Node II/117. Populating Next Right Pointers in Each Node II.cs
public class Solution {
    public Node Connect(Node root) {
        if (root == null) return null;
        
        Node curr = root;
        while (curr != null) {
            Node nextLevelStart = null;
            Node prev = null;
            
            while (curr != null) {
                // Handle left child
                if (curr.left != null) {
                    if (nextLevelStart == null) nextLevelStart = curr.left;
                    if (prev != null) prev.next = curr.left;
                    prev = curr.left;
                }
                
                // Handle right child
                if (curr.right != null) {
                    if (nextLevelStart == null) nextLevelStart = curr.right;
                    if (prev != null) prev.next = curr.right;
                    prev = curr.right;
                }
                
                curr = curr.next;
            }
            
            curr = nextLevelStart;
        }
        
        return root;
    }
}

File: 117. Populating Next Right Pointers in Each Node II/117. Populating Next Right Pointers in Each Node II.rb
def connect(root)
    return nil if root.nil?
    
    curr = root
    while curr
        next_level_start = nil
        prev = nil
        
        while curr
            # Handle left child
            if curr.left
                next_level_start ||= curr.left
                prev.next = curr.left if prev
                prev = curr.left
            end
            
            # Handle right child
            if curr.right
                next_level_start ||= curr.right
                prev.next = curr.right if prev
                prev = curr.right
            end
            
            curr = curr.next
        end
        
        curr = next_level_start
    end
    
    root
end